File: services/collector/hantawatch_collector/hpi.py
"""Python port of the HPI (Hanta Proximity Index) formula.

This intentionally mirrors `apps/web/src/lib/hpi.ts`. Both must stay in sync;
the front-end `/about#hpi` page is the source of truth for the *spec*, while
these two files are mechanical implementations.

If you adjust weights here, also adjust them in the TS file (and bump
documentation).
"""

from __future__ import annotations

from dataclasses import dataclass

# Weights — MUST equal those in hpi.ts
W_DISTANCE = 0.30
W_OFFICIAL = 0.25
W_SEROTYPE = 0.20
W_TRAVEL = 0.15
W_BASELINE = 0.10

# Serotype intrinsic risk scores (0–100). MUST match hpi.ts serotypeScore().
SEROTYPE_RISK = {
    "andes": 100,
    "sin_nombre": 85,
    "hantaan": 30,
    "seoul": 20,
    "puumala": 5,
    "other": 15,
}


@dataclass
class HpiInputs:
    distance_km: float
    official_risk_level: str  # 'low' | 'moderate' | 'high' | 'very_high'
    serotype_id: str
    travel_connectivity: str  # 'none' | 'indirect' | 'direct'
    baseline_deviation: str   # 'below' | 'normal' | 'elevated'


def _distance_score(km: float) -> float:
    if km > 10000:
        return 0.0
    if km > 3000:
        return 20.0
    if km > 500:
        return 50.0
    return 100.0


def _official_score(level: str) -> float:
    return {"low": 0.0, "moderate": 40.0, "high": 70.0, "very_high": 100.0}.get(level, 0.0)


def _travel_score(level: str) -> float:
    return {"none": 5.0, "indirect": 15.0, "direct": 40.0}.get(level, 5.0)


def _baseline_score(dev: str) -> float:
    return {"below": 0.0, "normal": 20.0, "elevated": 90.0}.get(dev, 20.0)


def _grade(total: float) -> tuple[str, str, str]:
    """Return (id, zh, color) for a given total. Matches GRADES in hpi.ts."""
    if total <= 20:
        return "low", "低关注", "#16a34a"
    if total <= 40:
        return "moderate", "一般关注", "#0891b2"
    if total <= 60:
        return "elevated", "中等关注", "#ca8a04"
    if total <= 80:
        return "high", "高度关注", "#ea580c"
    return "severe", "严重关注", "#dc2626"


def _js_round(x: float) -> int:
    """JavaScript-compatible Math.round: half away from zero.

    Python's built-in round() uses banker's rounding (half to even), which
    would diverge from the TS implementation on .5 boundaries (e.g.
    round(82.5) == 82 in Python but Math.round(82.5) == 83 in JS).
    """
    import math
    if x >= 0:
        return math.floor(x + 0.5)
    return -math.floor(-x + 0.5)


def calculate_hpi(inp: HpiInputs) -> dict:
    """Return an HpiResult-shaped dict matching `HpiResult` in shared types."""
    d_score = _distance_score(inp.distance_km)
    o_score = _official_score(inp.official_risk_level)
    s_score = SEROTYPE_RISK.get(inp.serotype_id, 15)
    t_score = _travel_score(inp.travel_connectivity)
    b_score = _baseline_score(inp.baseline_deviation)

    total = (
        d_score * W_DISTANCE
        + o_score * W_OFFICIAL
        + s_score * W_SEROTYPE
        + t_score * W_TRAVEL
        + b_score * W_BASELINE
    )
    total = max(0, min(100, _js_round(total)))
    grade_id, grade_zh, color = _grade(total)

    return {
        "total": total,
        "grade": grade_id,
        "gradeZh": grade_zh,
        "color": color,
        "factors": {
            "distance": {"km": int(inp.distance_km), "score": d_score, "weight": W_DISTANCE},
            "officialAssessment": {"level": inp.official_risk_level, "score": o_score, "weight": W_OFFICIAL},
            "serotypeRisk": {"serotypeId": inp.serotype_id, "score": s_score, "weight": W_SEROTYPE},
            "travelConnectivity": {"level": inp.travel_connectivity, "score": t_score, "weight": W_TRAVEL},
            "historicalBaseline": {"deviation": inp.baseline_deviation, "score": b_score, "weight": W_BASELINE},
        },
    }

File: apps/web/src/lib/hpi.ts
/**
 * HPI (Hanta Proximity Index) calculation engine.
 *
 * HPI = W₁×D + W₂×O + W₃×S + W₄×T + W₅×H
 *
 * All factor inputs and weight mappings are transparent and documented.
 */

export interface HpiInput {
  /** Distance from nearest active cluster to China border (km) */
  distanceKm: number;
  /** WHO/CDC/ChinaCDC official risk assessment level */
  officialRiskLevel: 'low' | 'moderate' | 'high' | 'very_high';
  /** Active serotype ID */
  serotypeId: string;
  /** Travel connectivity level from cluster region to China */
  travelConnectivity: 'none' | 'indirect' | 'direct';
  /** Deviation of China HFRS cases vs historical baseline */
  baselineDeviation: 'below' | 'normal' | 'elevated';
}

export interface HpiOutput {
  total: number;
  grade: 'low' | 'moderate' | 'elevated' | 'high' | 'severe';
  gradeZh: string;
  color: string;
  breakdown: {
    distance: { raw: number; score: number; weight: number; weighted: number };
    official: { raw: string; score: number; weight: number; weighted: number };
    serotype: { raw: string; score: number; weight: number; weighted: number };
    travel: { raw: string; score: number; weight: number; weighted: number };
    baseline: { raw: string; score: number; weight: number; weighted: number };
  };
}

// ---- Factor mappers (transparent, auditable) ----

/** Distance ring → score mapping */
function distanceScore(km: number): number {
  if (km > 10000) return 0;
  if (km > 3000) return 20;
  if (km > 500) return 50;
  return 100;
}

/** Official risk level → score mapping */
function officialScore(level: HpiInput['officialRiskLevel']): number {
  const map: Record<string, number> = {
    low: 0,
    moderate: 40,
    high: 70,
    very_high: 100,
  };
  return map[level] ?? 0;
}

/** Serotype intrinsic risk → score */
function serotypeScore(id: string): number {
  const map: Record<string, number> = {
    andes: 100,
    sin_nombre: 85,
    hantaan: 30,
    seoul: 20,
    puumala: 5,
    other: 15,
  };
  return map[id] ?? 15;
}

/** Travel connectivity → score */
function travelScore(level: HpiInput['travelConnectivity']): number {
  const map: Record<string, number> = {
    none: 5,
    indirect: 15,
    direct: 40,
  };
  return map[level] ?? 5;
}

/** Baseline deviation → score */
function baselineScore(dev: HpiInput['baselineDeviation']): number {
  const map: Record<string, number> = {
    below: 0,
    normal: 20,
    elevated: 90,
  };
  return map[dev] ?? 20;
}

// ---- Grade thresholds ----

interface Grade {
  id: HpiOutput['grade'];
  zh: string;
  color: string;
  min: number;
  max: number;
}

const GRADES: Grade[] = [
  { id: 'low', zh: '低关注', color: '#16a34a', min: 0, max: 20 },
  { id: 'moderate', zh: '一般关注', color: '#0891b2', min: 21, max: 40 },
  { id: 'elevated', zh: '中等关注', color: '#ca8a04', min: 41, max: 60 },
  { id: 'high', zh: '高度关注', color: '#ea580c', min: 61, max: 80 },
  { id: 'severe', zh: '严重关注', color: '#dc2626', min: 81, max: 100 },
];

// Weights
const W = {
  distance: 0.30,
  official: 0.25,
  serotype: 0.20,
  travel: 0.15,
  baseline: 0.10,
};

// ---- Public API ----

export function calculateHpi(input: HpiInput): HpiOutput {
  const dScore = distanceScore(input.distanceKm);
  const oScore = officialScore(input.officialRiskLevel);
  const sScore = serotypeScore(input.serotypeId);
  const tScore = travelScore(input.travelConnectivity);
  const hScore = baselineScore(input.baselineDeviation);

  const dWeighted = dScore * W.distance;
  const oWeighted = oScore * W.official;
  const sWeighted = sScore * W.serotype;
  const tWeighted = tScore * W.travel;
  const hWeighted = hScore * W.baseline;

  const total = Math.round(dWeighted + oWeighted + sWeighted + tWeighted + hWeighted);
  const clampedTotal = Math.max(0, Math.min(100, total));

  const grade = GRADES.find(g => clampedTotal >= g.min && clampedTotal <= g.max) ?? GRADES[0];

  return {
    total: clampedTotal,
    grade: grade.id,
    gradeZh: grade.zh,
    color: grade.color,
    breakdown: {
      distance: { raw: input.distanceKm, score: dScore, weight: W.distance, weighted: dWeighted },
      official: { raw: input.officialRiskLevel, score: oScore, weight: W.official, weighted: oWeighted },
      serotype: { raw: input.serotypeId, score: sScore, weight: W.serotype, weighted: sWeighted },
      travel: { raw: input.travelConnectivity, score: tScore, weight: W.travel, weighted: tWeighted },
      baseline: { raw: input.baselineDeviation, score: hScore, weight: W.baseline, weighted: hWeighted },
    },
  };
}

export { GRADES, W as HPI_WEIGHTS };
export { distanceScore, officialScore, serotypeScore, travelScore, baselineScore };
